Mathlib/Topology/CWComplex/Classical/Subcomplex.lean

```diff
@@ -0,0 +1,180 @@
+
+/-- A subcomplex is the union of its closed cells. -/
+lemma CWComplex.Subcomplex.union_closedCell [T2Space X] [CWComplex C] (E : Subcomplex C) :
+    ⋃ (n : ℕ) (j : E.I n), closedCell (C := C) n j = E :=
+  (empty_union _ ).symm.trans (RelCWComplex.Subcomplex.union_closedCell E)
+
+lemma RelCWComplex.Subcomplex.disjoint_openCell_subcomplex_of_not_mem [RelCWComplex C D]
+    (E : Subcomplex C) {n : ℕ} {i : cell C n} (h : i ∉ E.I n) : Disjoint (openCell n i) E := by
+  simp_rw [← union, disjoint_union_right, disjoint_iUnion_right]
+  exact ⟨disjointBase n i , fun _ _ ↦ disjoint_openCell_of_ne (by aesop)⟩
+
+open Classical in
+/-- A subcomplex is again a CW complex. -/
+@[simps]
+instance RelCWComplex.Subcomplex.instRelCWComplex [T2Space X] [RelCWComplex C D]
+    (E : Subcomplex C) : RelCWComplex E D where
+  cell n := E.I n
+  map n i := map (C := C) n i
+  source_eq n i := source_eq (C := C) n i
+  continuousOn n i := continuousOn (C := C) n i
+  continuousOn_symm n i := continuousOn_symm (C := C) n i
+  pairwiseDisjoint' := by
+    simp only [PairwiseDisjoint, Set.Pairwise, mem_univ, ne_eq, Function.onFun,
+      disjoint_iff_inter_eq_empty, true_implies, Sigma.forall, Subtype.forall]
+    intro n j _ m i _ eq
+    refine (disjoint_openCell_of_ne ?_).inter_eq
+    simp_all only [Sigma.mk.inj_iff, not_and, ne_eq]
+    intro a
+    subst a
+    simp_all
+  disjointBase' n i := disjointBase' (C := C) n i
+  mapsTo := by
+    intro n i
+    rcases cellFrontier_subset_finite_openCell (C := C) n i with ⟨J, hJ⟩
+    use fun m ↦ Finset.preimage (J m) Subtype.val (by simp [InjOn])
+    rw [mapsTo_iff_image_subset]
+    intro x hx
+    specialize hJ hx
+    simp_rw [iUnion_coe_set, mem_union, mem_iUnion, exists_prop, exists_and_right,
+      Decidable.or_iff_not_imp_left] at hJ ⊢
+    intro h
+    specialize hJ h
+    obtain ⟨m, hmn, j, hj, hxj⟩ := hJ
+    use m, hmn, j
+    refine ⟨?_, openCell_subset_closedCell _ _ hxj⟩
+    suffices j ∈ E.I m from ⟨this, by simp [hj]⟩
+    have : x ∈ (E : Set X) := by
+      rw [← union_closedCell]
+      right
+      refine mem_of_subset_of_mem ?_ hx
+      refine subset_iUnion_of_subset n (subset_iUnion_of_subset i ?_)
+      exact cellFrontier_subset_closedCell (C := C) n i
+    simp only [← E.union, iUnion_coe_set, mem_union, h, mem_iUnion, exists_prop,
+      false_or] at this
+    obtain ⟨l, o, hxo⟩ := this
+    suffices (⟨m, j⟩ : Σ n, cell C n) = ⟨l, ↑o⟩ by aesop
+    apply eq_of_not_disjoint_openCell
+    rw [not_disjoint_iff]
+    use x
+    exact ⟨hxj, hxo.2⟩
+  closed' A hA h := by
+    apply isClosed_of_isClosed_inter_openCell_or_isClosed_inter_closedCell
+      (subset_trans hA (subset_complex (C := C) E)) h.2
+    intro n _ j
+    by_cases hj : j ∈ E.I n
+    · exact Or.intro_right _ (h.1 n ⟨j, hj⟩)
+    left
+    suffices A ∩ openCell n j = ∅ from this ▸ isClosed_empty
+    rw [← disjoint_iff_inter_eq_empty]
+    exact (disjoint_openCell_subcomplex_of_not_mem E hj).symm.mono_left hA
```

Comment:
Do you think it would be worth proving a special case of `isClosed_of_isClosed_inter_openCell_or_isClosed_inter_closedCell` assuming directly that either `A` intersects the closed cell in a closed way, or `A` is disjoint from the open cell? Maybe that's a bit overkill, but the special case was already used in your PR defining subcomplexes.

Reply:
I'm not sure. My initial instinct was "oh I think it can be useful" but indeed I'm having troubles imagining an example where it would be more efficient... Is the general lemma coming from a textbook of did you coin it during the formalization?

Reply:
Then I would think removing it is fine. But this is not a strong opinion, I'm fine with it if you prefer to keep it.